cost_rules/app.py

```diff
@@ -0,0 +1,293 @@
+            ]
+          }
+        }
+      }
+
+    '''
+    rule = {
+        'Type': 'REGULAR',
+        'Value': label,
+    }
+
+    _accounts = {
+        'Dimensions': {
+            'Key': 'LINKED_ACCOUNT',
+            'Values': accounts,
+            'MatchOptions': [ 'EQUALS', ],
+        }
+    }
+
+    def _build_tag_absent(key):
+        tag = {}
+        tag['Key'] = key
+        tag['MatchOptions'] = [ 'ABSENT', ]
+        return {'Tags': tag}
+
+    _rule = {}
+    if tags:
+        _tags = [ _build_tag_absent(t) for t in tags ]
+        rule['Rule'] = { 'And': [ _accounts, ] + _tags }
+
+    else:
+        rule['Rule'] = _accounts
+
+    return rule
+
+
+def inherited_rules(tags):
+    '''
+    Fall back on directly using tag values
+    Example JSON
+
+
+  [{
+    "Type": "INHERITED_VALUE",
+    "InheritedValue": {
+      "DimensionName": "TAG",
+      "DimensionKey": "CostCenterOther"
+    }
+  },
+  {
+    "Type": "INHERITED_VALUE",
+    "InheritedValue": {
+      "DimensionName": "TAG",
+      "DimensionKey": "CostCenter"
+    }
+  }]
+    '''
+    rules = []
+
+    for t in tags:
+        rule = {
+            'Type': 'INHERITED_VALUE',
+            'InheritedValue': {
+                'DimensionName': 'TAG',
+                'DimensionKey': t,
+            }
+        }
+        rules.append(rule)
+
+    return rules
+
+def rule_generator(params):
+    '''
+        Input:
+
+        InheritedValues:
+          TagOrder:
+            - CostCenterOther
+            - CostCenter
+          RulePosition: Last
+        RegularValues:
+          - Value: Category One
+            Accounts:
+              - 123abc
+              - 456xyz
+            TagNames:
+              - CostCenter
+              - CostCenterOther
+            TagEndsWith:
+              - 123400
+          - Value: Category Two
+            ...
+    '''
+    # Rule order matters, first match wins and rules should
+    # generally be ordered most-common to least-common
+    rules = []
+
+    # First get the inherited tag info
+    it_position = 'Last'  # either 'First' or 'Last'
+    it_rules = []
+    if 'InheritedValues' in params:
+        info = params['InheritedValues']
+
+        if 'RulePosition' in info:
+            it_position = info['RulePosition']
+
+        if 'TagOrder' in info:
+            order = info['TagOrder']
+            it_rules = inherited_rules(order)
+
+    # add our inherited tag rules if their position is First
+    if it_rules and it_position == 'First':
+        rules.extend(it_rules)
+
+    # iterate over our labels, building the rules for each
+    if 'RegularValues' in params:
+        for info in params['RegularValues']:
+
+            # First add tag suffix rules
+            tags = None
+            if 'TagNames' in info:
+                tags = info['TagNames']
+                if 'TagEndsWith' in info:
+                    rules.extend(tag_rules(info['Value'], tags, info['TagEndsWith'], 'ENDS_WITH'))
```

Comment:
`Value` is required so maybe some error handling for that?

Reply:
also added a KeyError here

Reply:
add a test?